### ISSUE TITLE
HTTP multipart/XRootD vector_read for remote baskets

### ISSUE THREAD

**Issue by jpivarski** [scikit-hep/uproot#393]
The chunked sources (`HTTPSource` and `XRootDSource`) currently have a fixed-chunksize cache. Since the base directory and TTree metadata of most files is at the end of the file, this means that `uproot.open("http://...")` consists of two sequential remote reads: the beginning of the file (`TFile` header and all `TStreamerInfos`) and the end of the file, and the `file["tree"]` request probably hits the latter. This is a good argument for making the chunksize large (1 MB, PR #392), so that it is not likely to be more than two such reads.

However, when a user requests a single or a subset of branches from the TTree, a large chunksize might require too much data downloading because 1 MB is coarser than most basket sizes. Ideally, we want to send an HTTP multipart or XRootD vector_read of _only_ the bytes corresponding to the basket data. They are known before the read:

```python
branch._fBasketSeek[:branch.numbaskets]    # start of each basket in bytes
branch._fBasketBytes[:branch.numbaskets]   # size of each basket in bytes
```

The main chunked source cache relies on the fact that the chunksize is fixed, so that finding the appropriate chunk is a constant-time operation (it is performed many times, for every set of contiguous, fixed-size fields and for 2-3 times for every variable-size string). These baskets would be variable-sized, so they should be a _second cache_, in addition to the main cache. The second cache would have to maintain a list of start positions and sizes, with log-N lookup time via Python's `bisect` or Numpy's `searchsorted`.

Requesting data from baskets would have to be called in a different way for baskets than for every other case. This is not a major problem, since such access is already abstracted through `Cursor` and `Source`: a new method could be added to `Source`, like `Source.data` but going through the second cache (in chunked sources; no difference in other sources) and a new method could be added to `Cursor` for getting an array that is only used in basket reading.

The HTTP multipart or XRootD vector_read call could be added to `HTTPSource.preload` and `XRootDSource.preload` instead of the one-request-per-task parallel preloader that's in there now. The hook exists, but it will need to be expanded from knowing just the start of each seek point of interest to knowing the full ranges (starts and stops or starts and sizes).

Note that a `Cursor`'s request for fixed-fields and strings in a basket's `TKey` header must also hit the second cache—this is not just about passing raw basket bytes to Numpy. See the `TTreeMethods._threadsafe_iterate_keys` method, which loads half (`complete=False`) or all (`complete=True`) of each basket `TKey`. All of the basket `TKeys` must be completely read and interpreted before beginning the basket data compression and conversion into arrays because important information for the `Interpretation` can only be discovered this way, such as the number of items contained in all the nested arrays of a jagged array branch. This first pass over the multipart data must not be allowed to cause a cache eviction, or arrays might have to be downloaded twice: once before reading the `TKeys` and again afterward. It is not possible to make a multipart fetch for only the `TKey` headers, small though they are, because their exact size is not known from TBranch metadata.

**Comment by chrisburr** [comment]
XRootD can be tested using files in:
```bash
xrdfs root://eospublic.cern.ch/ ls -l /eos/opendata/lhcb/AntimatterMatters2017/data
-r-- 2017-03-07 13:53:05   666484974 /eos/opendata/lhcb/AntimatterMatters2017/data/B2HHH_MagnetDown.root
-r-- 2017-03-07 13:53:08   444723234 /eos/opendata/lhcb/AntimatterMatters2017/data/B2HHH_MagnetUp.root
-r-- 2017-03-07 13:53:08     2272072 /eos/opendata/lhcb/AntimatterMatters2017/data/PhaseSpaceSimulation.root
```
![IMG_0278](https://user-images.githubusercontent.com/5220533/68186528-49fe0c80-fff4-11e9-9e5a-26cf25b64062.jpg)

**Comment by chrisburr** [comment]
It turns out it is possible to request the last `50` bytes of a file with HTTP using `Range: bytes=0-50, -50`.

Do you have opinions about depending on https://github.com/requests/toolbelt for decoding the multipart HTTP responses? I think it's pure python so it shouldn't cause any trouble.

**Comment by jpivarski** [comment]
I prefer to avoid dependencies whenever possible—like there's a cost term associated with it that has to be justified with an equally compelling benefit. If it's hard to parse the multipart response, then that's good justification. If the dependency is pure Python and/or very common, those features lower the cost. Note that we strictly depend on `requests`, which has some nice features for HTTP.

**Comment by jpivarski** [comment]
@chrisburr Reassign this to someone else if you find someone or close it if it's time to give up. I've only assigned this to you to make it easier for me to scan over the issues that are still waiting for me to triage. (Maybe I should introduce an "already triaged" label...).

**Comment by jpivarski** [comment]
I believe you have implemented this in scikit-hep/uproot4#4. Thanks!
